File: LogicAndTrick.WikiCodeParser/ParseData.cs
using System;
using System.Collections.Generic;

namespace LogicAndTrick.WikiCodeParser
{
    public class ParseData
    {
        private readonly Dictionary<string, object> _values;

        public ParseData()
        {
            _values = new Dictionary<string, object>();
        }

        public T Get<T>(string key, Func<T> defaultValue)
        {
            if (_values.ContainsKey(key) && _values[key] is T val) return val;
            var v = defaultValue();
            _values[key] = v;
            return v;
        }

        public void Set<T>(string key, T value)
        {
            _values[key] = value;
        }
    }
}

File: php/src/ParseData.php
<?php

namespace LogicAndTrick\WikiCodeParser;

class ParseData
{
    private array $values;

    public function __construct() {
        $this->values = [];
    }

    public function &Get(string $key, callable $defaultValue) : mixed {
        if (array_key_exists($key, $this->values)) return $this->values[$key];
        $v = $defaultValue();
        $this->values[$key] = $v;
        return $v;
    }

    public function Set(string $key, mixed &$value) {
        $this->values[$key] = &$value;
    }
}
